ta/pkcs11/src/pkcs11_attributes.c

```diff
@@ -1303,3 +1338,223 @@ bool attribute_is_exportable(struct pkcs11_attribute_head *req_attr,
+	case PKCS11_CKA_SUBJECT:
+	case PKCS11_CKA_ENCRYPT:
+	case PKCS11_CKA_VERIFY:
+	case PKCS11_CKA_VERIFY_RECOVER:
+	case PKCS11_CKA_WRAP:
+		return true;
+	case PKCS11_CKA_TRUSTED:
+		/* Change in CKS_TRUSTED can only be done by SO */
+		return pkcs11_session_is_so(session);
+	default:
+		return false;
+	}
+}
+
+static bool attr_is_modifiable_private_key(struct pkcs11_attribute_head *attr,
+					   struct pkcs11_session *sess __unused,
+					   struct pkcs11_object *obj)
+{
+	switch (attr->id) {
+	case PKCS11_CKA_SUBJECT:
+	case PKCS11_CKA_DECRYPT:
+	case PKCS11_CKA_SIGN:
+	case PKCS11_CKA_SIGN_RECOVER:
+	case PKCS11_CKA_UNWRAP:
+	/*
+	 * TBD: Revisit if we don't support PKCS11_CKA_PUBLIC_KEY_INFO
+	 * Specification mentions that if this attribute is
+	 * supplied as part of a template for C_CreateObject, C_CopyObject or
+	 * C_SetAttributeValue for a private key, the token MUST verify
+	 * correspondence between the private key data and the public key data
+	 * as supplied in CKA_PUBLIC_KEY_INFO. This needs to be
+	 * taken care of when this object type will be implemented
+	 */
+	case PKCS11_CKA_PUBLIC_KEY_INFO:
+		return true;
+	/* Can't be modified once set to CK_FALSE - 12 in Table 10 */
+	case PKCS11_CKA_EXTRACTABLE:
+		return get_bool(obj->attributes, PKCS11_CKA_EXTRACTABLE);
+	/* Can't be modified once set to CK_TRUE - 11 in Table 10 */
+	case PKCS11_CKA_SENSITIVE:
+		return !get_bool(obj->attributes, PKCS11_CKA_SENSITIVE);
+	case PKCS11_CKA_WRAP_WITH_TRUSTED:
+		return !get_bool(obj->attributes, PKCS11_CKA_WRAP_WITH_TRUSTED);
+	default:
+		return false;
+	}
+}
+
+static bool attribute_is_modifiable(struct pkcs11_session *session,
+				    struct pkcs11_attribute_head *req_attr,
+				    struct pkcs11_object *obj,
+				    enum pkcs11_class_id class,
+				    enum processing_func function)
+{
+	/* Check modifiable attributes common to any object */
+	switch (req_attr->id) {
+	case PKCS11_CKA_LABEL:
+		return true;
+	case PKCS11_CKA_TOKEN:
+	case PKCS11_CKA_PRIVATE:
+	case PKCS11_CKA_MODIFIABLE:
+	case PKCS11_CKA_DESTROYABLE:
+		return function == PKCS11_FUNCTION_COPY;
+	case PKCS11_CKA_COPYABLE:
+		/*
+		 * TBD - Specification mentions that if the attribute value
+		 * is false it can't be set to true. Reading this it looks
+		 * like it should be possible to modify this attribute by
+		 * C_SetAttributeValue() but this is not marked as modifiable
+		 * in Table 10 neither it is mentioned that it can be modified
+		 * anywhere else. So currently assuming we can modify it.
+		 */
+		if (function == PKCS11_FUNCTION_MODIFY)
+			/* Only allow change from CK_TRUE -> CK_FALSE */
+			return get_bool(obj->attributes, PKCS11_CKA_COPYABLE);
+		else
+			return false;
+	default:
+		break;
+	}
+
+	/* Attribute checking based on class type */
+	switch (class) {
+	case PKCS11_CKO_SECRET_KEY:
+	case PKCS11_CKO_PUBLIC_KEY:
+	case PKCS11_CKO_PRIVATE_KEY:
+		if (attr_is_modifiable_any_key(req_attr))
+			return true;
+		if (class == PKCS11_CKO_SECRET_KEY &&
```

Comment:
perhaps it would be just clearer if each switch clause would have any key and then specific class style function return value like:
```c
case PKCS11_CKO_SECRET_KEY:
	if (attr_is_modifiable_any_key(req_attr))
		return true;
	return attr_is_modifiable_secret_key(req_attr, session, obj));
```

Reply:
I am ok with either ways. In rest of the code, I have usually seen it being done this way so I followed this convention. If @etienne-lms also agrees, I can change this.